### PR TITLE
SONAR-6572 ws metrics/update update an existing custom metric

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+      updateMetricInDb(dbSession, metricInDb, metricTemplate);
+      JsonWriter json = response.newJsonWriter();
+      writeMetric(json, metricInDb);
+      json.close();
+    } finally {
+      MyBatis.closeQuietly(dbSession);
+    }
+  }
+
+  private MetricDto newMetricTemplate(Request request) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java]: MAJOR Make "newMetricTemplate" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+  private boolean isMetricNonCustom(@Nullable MetricDto metricInDb) {
+    return !metricInDb.isUserManaged();
+  }
+
+  private boolean isMetricDisabled(@Nullable MetricDto metricInDb) {
+    return !metricInDb.isEnabled();
+  }
+
+  private boolean isMetricFoundInDb(@Nullable MetricDto metricInDb) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java]: MAJOR Make "isMetricFoundInDb" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    if (haveMetricTypeChanged(metricInDb, template)) {
+      List<CustomMeasureDto> customMeasures = dbClient.customMeasureDao().selectByMetricId(dbSession, metricInDb.getId());
+      if (haveAssociatedCustomMeasures(customMeasures)) {
+        throw new ServerException(HttpURLConnection.HTTP_CONFLICT, String.format("You're trying to change the type '%s' while there are associated custom measures.",
+          metricInDb.getValueType()));
+      }
+    }
+  }
+
+  private void writeMetric(JsonWriter json, MetricDto metric) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java]: MAJOR Make "writeMetric" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    json.prop(FIELD_DESCRIPTION, metric.getDescription());
+    json.endObject();
+  }
+
+  private boolean isMetricNonCustom(@Nullable MetricDto metricInDb) {
+    return !metricInDb.isUserManaged();
+  }
+
+  private boolean isMetricDisabled(@Nullable MetricDto metricInDb) {
+    return !metricInDb.isEnabled();
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java]: CRITICAL metricInDb must be non-null but is marked as nullable [rule](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE)

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    json.prop(FIELD_DOMAIN, metric.getDomain());
+    json.prop(FIELD_DESCRIPTION, metric.getDescription());
+    json.endObject();
+  }
+
+  private boolean isMetricNonCustom(@Nullable MetricDto metricInDb) {
+    return !metricInDb.isUserManaged();
+  }
+
+  private boolean isMetricDisabled(@Nullable MetricDto metricInDb) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java]: MAJOR Make "isMetricDisabled" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+  private boolean isMetricFoundInDb(@Nullable MetricDto metricInDb) {
+    return metricInDb == null;
+  }
+
+  private boolean haveAssociatedCustomMeasures(List<CustomMeasureDto> customMeasures) {
+    return !customMeasures.isEmpty();
+  }
+
+  private boolean haveMetricTypeChanged(@Nullable MetricDto metricInDb, MetricDto template) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java]: MAJOR Make "haveMetricTypeChanged" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    json.prop(FIELD_KEY, metric.getKey());
+    json.prop(FIELD_TYPE, metric.getValueType());
+    json.prop(FIELD_NAME, metric.getShortName());
+    json.prop(FIELD_DOMAIN, metric.getDomain());
+    json.prop(FIELD_DESCRIPTION, metric.getDescription());
+    json.endObject();
+  }
+
+  private boolean isMetricNonCustom(@Nullable MetricDto metricInDb) {
+    return !metricInDb.isUserManaged();
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java]: CRITICAL metricInDb must be non-null but is marked as nullable [rule](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE)

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+  private boolean isMetricFoundInDb(@Nullable MetricDto metricInDb) {
+    return metricInDb == null;
+  }
+
+  private boolean haveAssociatedCustomMeasures(List<CustomMeasureDto> customMeasures) {
+    return !customMeasures.isEmpty();
+  }
+
+  private boolean haveMetricTypeChanged(@Nullable MetricDto metricInDb, MetricDto template) {
+    return !metricInDb.getValueType().equals(template.getValueType()) && template.getValueType() != null;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java]: CRITICAL metricInDb must be non-null but is marked as nullable [rule](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE)

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+    json.prop(FIELD_ID, String.valueOf(metric.getId()));
+    json.prop(FIELD_KEY, metric.getKey());
+    json.prop(FIELD_TYPE, metric.getValueType());
+    json.prop(FIELD_NAME, metric.getShortName());
+    json.prop(FIELD_DOMAIN, metric.getDomain());
+    json.prop(FIELD_DESCRIPTION, metric.getDescription());
+    json.endObject();
+  }
+
+  private boolean isMetricNonCustom(@Nullable MetricDto metricInDb) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java]: MAJOR Make "isMetricNonCustom" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+  private boolean isMetricDisabled(@Nullable MetricDto metricInDb) {
+    return !metricInDb.isEnabled();
+  }
+
+  private boolean isMetricFoundInDb(@Nullable MetricDto metricInDb) {
+    return metricInDb == null;
+  }
+
+  private boolean haveAssociatedCustomMeasures(List<CustomMeasureDto> customMeasures) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/UpdateAction.java]: MAJOR Make "haveAssociatedCustomMeasures" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 10 new issues:
* 3 critical
* 7 major

File: server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java [hunk 11]
```diff
@@ -188,7 +188,7 @@ private void checkMetricInDbAndTemplate(@Nullable MetricDto metricInDb, MetricDt
 
   private void writeMetric(JsonWriter json, MetricDto metric) {
     json.beginObject();
-    json.prop(FIELD_ID, metric.getId());
+    json.prop(FIELD_ID, String.valueOf(metric.getId()));
```
Comment (jblievremont) [inline server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java]: metric.getId().toString()?

File: server/sonar-server/src/test/java/org/sonar/server/metric/ws/CreateActionTest.java [hunk 12]
```diff
@@ -131,7 +131,7 @@ public void return_metric_with_id() throws Exception {
       .execute();
 
     result.assertJson(getClass(), "metric.json");
-    result.outputAsString().matches("\"id\"\\s*:\\s*\"\\w+\"");
+    assertThat(result.outputAsString()).matches(".*\"id\"\\s*:\\s*\"\\w+\".*");
```
Comment (jblievremont) [inline server/sonar-server/src/test/java/org/sonar/server/metric/ws/CreateActionTest.java]: assertThat(result.outputAsString()).contains("id")?

File: server/sonar-server/src/test/java/org/sonar/server/metric/ws/UpdateActionTest.java [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.sonar.server.metric.ws.UpdateAction.PARAM_DESCRIPTION;
+import static org.sonar.server.metric.ws.UpdateAction.PARAM_DOMAIN;
+import static org.sonar.server.metric.ws.UpdateAction.PARAM_ID;
+import static org.sonar.server.metric.ws.UpdateAction.PARAM_KEY;
+import static org.sonar.server.metric.ws.UpdateAction.PARAM_NAME;
+import static org.sonar.server.metric.ws.UpdateAction.PARAM_TYPE;
+
+@Category(DbTests.class)
+public class UpdateActionTest {
```
Comment (jblievremont) [inline server/sonar-server/src/test/java/org/sonar/server/metric/ws/UpdateActionTest.java]: Please add a test for "missing permission"

**Comment by jblievremont** [comment]
Very minor feedback, then LGTM!

File: server/sonar-server/src/test/java/org/sonar/server/metric/ws/CreateActionTest.java (hunk 12 above)
Comment (teryk) [inline server/sonar-server/src/test/java/org/sonar/server/metric/ws/CreateActionTest.java]: nope, I'm checking "id":"xxx" where xxx is a number of alpha numeric characters

**Pull request merged**